Adding user-provided torsion moment of inertia #10313

sveerara opened the pull request:
Usually in beams the torsion moment of inertia (Ix - along the axis of the beam) is calculated as the sum of the two bending moments of inertia (Iy+Iz). However, in some scenarios Ix may be different. This PR allows the user to specify Ix. It defaults to Iy + Iz if the parameter is not specified by the user. 

Also adds quasi-static torsion tests.
refs #10313

sveerara committed:
Adding user-provided torsion moment of inertia #10313

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/201086/) on 3f7774d wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11683/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11683/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5a2f370d09c9355f4349998d80817e81cf0234b4`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/201096/) on d42766a wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11683/site)

This comment will be updated on new commits.

sveerara commented:
@hoffwm, @cbolisetti and @bwspenc, please look through this PR when you have time. Thanks!

permcody commented on modules/tensor_mechanics/src/kernels/InertialForceBeam.C:
```diff
@@ -80,6 +82,7 @@ InertialForceBeam::InertialForceBeam(const InputParameters & parameters)
     _area(coupledValue("area")),
     _Ay(coupledValue("Ay")),
     _Az(coupledValue("Az")),
+    _Ix(isParamValid("Ix") ? coupledValue("Ix") : _zero),
```
This is good, but I noticed that you are checking for the validity of the other parameters even though you are using them to get coupling components. This is a bad idea. You should either make them required, provide a default (my favorite), or check each of them before accessing their values.

permcody commented:
Ping @bwspenc - This one has been sitting for awhile.

sveerara commented:
@permcody & @bwspenc, I need to address Cody's earlier comments on the PR and I also made one more change to the input parameters list (user can provide z_orientation instead of y_orientation of the beam). I will add those changes as well later today so that Ben can review all the changes simultaneously. Thanks!

sveerara committed:
Adding nodal mass using file, nodal gravity and changes to line element action #10313

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/216530/) on 3300a62 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11683/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11683/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8ac020c40a018c04fc4c7cec5068109732035c90`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/216586/) on 1731eae wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11683/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11683/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8ac020c40a018c04fc4c7cec5068109732035c90`

sveerara commented:
@permcody, I have addressed your comments on this PR. 

I am not sure I understand why the documentation test is failing. I have added a new object called NodalGravity.md and test is failing because there is an existing object called Gravity.md in the TensorMechanics module. I don't see why this should be a problem.

@bwspenc, @cbolisetti and @hoffwm: To input nodal mass for multiple nodes easily, I have added the capability to provide the nodal mass information through a csv file along with the nodal positions. I have added this an optional parameter in NodalTranslationalInertia and NodalGravity NodalKernels. 

Also, I have made few changes to the LineElementAction. Earlier the velocity and acceleration auxvariables and the corresponding auxkernels were created only if the inertial force kernels or nodal kernels are also set up by the action. I have added another option `add_dynamic_variables` which just adds the velocity and acceleration auxvariables and the corresponding auxkernels for all the beam element blocks set up by the action. 

I have also corrected a bug in the action where some of the parameters such as `dynamic_nodal_translational_inertia` were being ignored if provided in the common action block.

bwspenc commented on modules/tensor_mechanics/doc/content/source/nodalkernels/NodalGravity.md:
```diff
@@ -0,0 +1,13 @@
+# Description
+
+This NodalKernel computes the gravitational force proportional to nodal mass in the coordinate direction corresponding to the assigned variable. A constant gravitational acceleration can be provided using `gravity_value` parameter and an acceleration varying in space and time can be provided using the `function` parameter. A constant mass for all the nodes in the given boundary can be provided us ...
+
```
Add some explanation about the formatting of the nodal mass file (maybe a simple example provided via a link to one of the test cases would be helpful).

bwspenc commented on modules/tensor_mechanics/src/nodalkernels/NodalGravity.C:
```diff
@@ -0,0 +1,112 @@
+  if (isParamValid("mass"))
+    mass = _mass;
+  else
+    mass = _node_id_to_mass[_current_node->id()];
```
You should probably check to make sure that there is an entry in the map for the current node id. If it's not there, this will just insert a zero into the map for that node, which isn't what you would want to do. You should error out in that case.

bwspenc commented on modules/tensor_mechanics/src/nodalkernels/NodalTranslationalInertia.C:
```diff
@@ -68,6 +124,12 @@ NodalTranslationalInertia::computeQpResidual()
+    if (isParamValid("mass"))
+      mass = _mass;
+    else
+      mass = _node_id_to_mass[_current_node->id()];
```
Same thing here about checking if the entry is in the map.

bwspenc reviewed (changes requested):
Just a few small things.  We're going to need to sort out that documentation build failure before we can merge this.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/223995/) on 44fae4c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](http://mooseframework.inl.gov/docs/PRs/11683/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s http://mooseframework.inl.gov/docs/PRs/11683/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format af8ec8e6cf0e693710321ab79c1e71a454fff4be`

sveerara commented:
@aeslaughter, can you please me fix the error with documentation? 

I have added a .md file for NodalGravity NodalKernel in TensorMechanics. There is another .md file for Gravity kernel in TensorMechanics. I am getting the following error:

`ERROR:The 'maxcount' was set to 1 but 2 nodes were found for the name 'Gravity.md'. `

I am not sure why the system thinks NodalGravity.md and Gravity.md are the same file.

sveerara commented:
@permcody, do I have to edit the file name (NodalGravity.md) or the first line in the .md file which is currently  
`# Nodal Gravity`

I tried editing the first line of the .md file to `# Nodal-Gravity` and `#\bNodal Gravity\b` and I still get the error. I am not sure if I am using the word boundary marker correctly. I haven't tried changing the file name yet. Since the file name is connected to the .C file name, I am not sure if I should be changing it.

permcody commented:
> @permcody, do I have to edit the file name (NodalGravity.md) or the first line in the .md file... ?

There error is coming from your links (Where you are referring to the `*.md` files and expecting MooseDocs to create HTML links). In this case it's in your `tests` file. You may want to run MooseDocs locally so you can check your changes and your rendering in real time (e.g. ./moosedoc.py build --serve)

sveerara commented:
Thanks @permcody. Adding a / before Gravity.md in the Porousflow documentation fixed the issue.

@bwspenc, I have addressed your earlier comments on the PR. Please review the PR when you get a chance.
